rust/src/pop3/pop3.rs

```diff
@@ -0,0 +1,767 @@
+    }
+}
+
+impl POP3State {
+    pub fn new() -> Self {
+        Self {
+            state_data: AppLayerStateData::new(),
+            tx_id: 0,
+            transactions: VecDeque::new(),
+            request_gap: false,
+            response_gap: false,
+        }
+    }
+
+    // Free a transaction by ID.
+    fn free_tx(&mut self, tx_id: u64) {
+        if let Some(index) = self.transactions.iter().position(|tx| tx.id() == tx_id + 1) {
+            self.transactions.remove(index);
+        }
+    }
+
+    pub fn get_tx(&self, tx_id: u64) -> Option<&POP3Transaction> {
+        self.transactions.iter().find(|tx| tx.id() == tx_id + 1)
+    }
+
+    pub fn get_tx_mut(&mut self, tx_id: u64) -> Option<&mut POP3Transaction> {
+        self.transactions.iter_mut().find(|tx| tx.id() == tx_id + 1)
+    }
+
+    fn new_tx(&mut self) -> POP3Transaction {
+        self.tx_id += 1;
+        POP3Transaction::new(self.tx_id)
+    }
+
+    fn find_request(&mut self) -> Option<&mut POP3Transaction> {
+        self.transactions
+            .iter_mut()
+            .find(|tx| tx.response.is_none())
+    }
+
+    fn parse_request(&mut self, input: &[u8]) -> AppLayerResult {
+        // We're not interested in empty requests.
+        if input.is_empty() {
+            return AppLayerResult::ok();
+        }
+
+        // If there was gap, check we can sync up again.
+        if self.request_gap {
+            unsafe {
+                if probe(input) != ALPROTO_POP3 {
+                    // The parser now needs to decide what to do as we are not in sync.
+                    // For this pop3, we'll just try again next time.
+                    return AppLayerResult::ok();
+                }
+            }
+
+            // It looks like we're in sync with a message header, clear gap
+            // state and keep parsing.
+            self.request_gap = false;
+        }
+
+        let mut start = input;
+        while !start.is_empty() {
+            match POP3_PARSER.parse(start, Direction::ToServer) {
+                Ok((rem, Some(msg))) => {
+                    if let InnerMessage::Command(request) = msg.inner {
+                        let mut tx = self.new_tx();
+                        tx.error_flags |= msg.error_flags;
+                        tx.request = Some(request);
+                        self.transactions.push_back(tx);
+                    }
+
+                    start = rem;
+                }
+                Ok((rem, None)) => {
+                    // Not enough data. This parser doesn't give us a good indication
+                    // of how much data is missing so just ask for one more byte so the
+                    // parse is called as soon as more data is received.
+
+                    let consumed = input.len() - rem.len();
+                    let needed = rem.len() + 1;
+                    return AppLayerResult::incomplete(consumed as u32, needed as u32);
+                }
+                Err(SawpError {
```

Comment:
how does this relate to the condition above? Couldn't the parser return a Needed::Size that includes the +1, or do you consider that to be suricata specific logic?

Reply:
sawp considers these different, the above condition could have progressed the buffer (some rem != start), where this condition didn't.